src/_posts/2019-12-30-backstopjs-tutorial.md

```diff
@@ -0,0 +1,257 @@
+
+## How
+
+Setting up BackstopJS isn't difficult. But I'm going to make it more difficult because we want to set it up in a way that's repeatable (ie it always has the same result for the same version of the software) and automated. You don't want your visual regression tests to be "flaky", to pass some times and not others. It's worth putting extra effort in to get this right, otherwise they may be less than useful. And you want them to be automated because that way you can trust yourself and others not to forget to run the test or update the snapshots, and you can use them as checks for Dependabot's auto-merge.
+
+Ultimately, your workflow should look like:
+
+1. You push some horribly misguided code, introducing an unintended visual change.
+2. Your CI runs the visual regression tests and they fail, turning your build red.
+3. You look at BackstopJS's test report from your CI, and spot the problem.
+4. You decide whether this change is acceptable, and either update the snapshots to this change or fix the code that caused this change.
+
+### Docker
+
+The first step to creating a repeatable visual regression test is to run the tests on the same platform every time. Otherwise, you're in for trouble. Small changes in things like font rendering between operating systems can prevent the reference snapshots generated on your local machine from matching the test snapshots generated on your CI server. And if your CI server has multiple test runners on different platforms you've got even more unpredictability on your hands. To get around issues like these we use Docker containers via Docker Compose. This guaruntees the same platform for every test run. This approach also has the advantage that you don't need to install BackstopJS locally and end up with different versions of it on each developer's machine; instead you have one consistent version, and Dependabot can keep it up-to-date. The disadvantage is of course that it's slower.
+
+Add the following entry to your root `docker-compose.yml` (create one if necessary):
+
+```(docker-compose)
+  visual_regression_tests:
+    image: backstopjs/backstopjs:4.4.2
+    volumes:
+      - ./test/visual:/src
+```
+
+This describes a `visual_regression_tests` Docker container using the official `backstopjs` image, version `4.4.2`. The version can be left out, but it's important that it be there for repeatability. Let Dependabot keep it up-to-date. Of course, you should copy the latest version number from [the BackstopJS Docker image releases](https://hub.docker.com/r/backstopjs/backstopjs/builds) and use that; `4.4.2` may be outdated at time of reading. If you're using a Docker container for your website as well you should add a `link` entry to that container.
```

Comment:
Does dependabot update docker-compose? We have not yet configured it to do that.

Reply:
Good point, I knew it could update Dockerfiles so assumed it could update Docker Compose but looks like it can't yet: https://github.com/dependabot/feedback/issues/82

Reply:
Fixed in 69a1ee67